[ONNX] De-duplicate initializers (#68202)

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69550
* #69548
* __->__ #69547
* #72663
* #68491

ScriptModule export introduces duplicated ONNX initializers for shared weights, unnecessarily increases ONNX model size. This PR de-duplicates ONNX initializers for model exported in eval mode, by checking if the underlying tensors share the same `data_ptr`, `strides` and `sizes`.

Co-authored-by: BowenBao <bowbao@microsoft.com>

Differential Revision: [D32994271](https://our.internmc.facebook.com/intern/diff/D32994271)

BowenBao committed:
[ONNX] De-duplicate initializers (#68202)

ScriptModule export introduces duplicated ONNX initializers for shared weights, unnecessarily increases ONNX model size. This PR de-duplicates ONNX initializers for model exported in eval mode, by checking if the underlying tensors share the same `data_ptr`, `strides` and `sizes`.

Co-authored-by: BowenBao <bowbao@microsoft.com>

[ghstack-poisoned]

msaroufim commented:
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32994271).

malfet commented on torch/csrc/jit/passes/onnx/helper.h:
```diff
@@ -14,6 +14,8 @@ static const int OPSET_VERSION_10 = 10;
 static const int OPSET_VERSION_11 = 11;
 static const int OPSET_VERSION_12 = 12;
 static const int OPSET_VERSION_13 = 13;
+static const int OPSET_VERSION_14 = 14;
+static const int OPSET_VERSION_15 = 15;
```
BE question: shouldn't it be enum?
```suggestion
enum {
  OPSET_VERSION_11 = 11,
  OPSET_VERSION_12,
  OPSET_VERSION_13,
  OPSET_VERSION_14,
  OPSET_VERSION_15,
};
```

malfet commented on torch/onnx/__init__.py:
```diff
@@ -10,7 +10,7 @@
 producer_name = "pytorch"
 producer_version = _C._onnx.PRODUCER_VERSION
-constant_folding_opset_versions = [9, 10, 11, 12, 13, 14]
+constant_folding_opset_versions = [9, 10, 11, 12, 13, 14, 15]
```
Could this list ever be contiguous? If not, lets ask python to generate it
```suggestion
constant_folding_opset_versions = list(range(9, 16))
```

BowenBao committed:
Update on "[ONNX] De-duplicate initializers (#68202)"


ScriptModule export introduces duplicated ONNX initializers for shared weights, unnecessarily increases ONNX model size. This PR de-duplicates ONNX initializers for model exported in eval mode, by checking if the underlying tensors share the same `data_ptr`, `strides` and `sizes`.

Co-authored-by: BowenBao <bowbaomicrosoft.com>

Differential Revision: [D32994271](https://our.internmc.facebook.com/intern/diff/D32994271)

[ghstack-poisoned]

msaroufim commented:
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32994271).

BowenBao committed:
Update on "[ONNX] De-duplicate initializers (#68202)"


ScriptModule export introduces duplicated ONNX initializers for shared weights, unnecessarily increases ONNX model size. This PR de-duplicates ONNX initializers for model exported in eval mode, by checking if the underlying tensors share the same `data_ptr`, `strides` and `sizes`.

Co-authored-by: BowenBao <bowbaomicrosoft.com>

Differential Revision: [D32994271](https://our.internmc.facebook.com/intern/diff/D32994271)

[ghstack-poisoned]

msaroufim commented:
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32994271).

msaroufim commented:
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32994271).

BowenBao committed:
Update on "[ONNX] De-duplicate initializers (#68202)"


ScriptModule export introduces duplicated ONNX initializers for shared weights, unnecessarily increases ONNX model size. This PR de-duplicates ONNX initializers for model exported in eval mode, by checking if the underlying tensors share the same `data_ptr`, `strides` and `sizes`.

Co-authored-by: BowenBao <bowbaomicrosoft.com>

Differential Revision: [D32994271](https://our.internmc.facebook.com/intern/diff/D32994271)

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32994271).
